components/image.js

```diff
@@ -132,3 +136,95 @@ export default function ZoomableImage ({ src, srcSet, ...props }) {
 }
+
+export const ImageUpload = forwardRef(({ children, className, onSelect, onSuccess, onError }, ref) => {
+  const toaster = useToast()
+
+  const [getSignedPOST] = useMutation(
+    gql`
+      mutation getSignedPOST($type: String!, $size: Int!, $width: Int!, $height: Int!) {
+        getSignedPOST(type: $type, size: $size, width: $width, height: $height) {
+          url
+          fields
+        }
+      }`)
+
+  const s3Upload = useCallback(file => {
+    const img = new window.Image()
+    img.src = window.URL.createObjectURL(file)
+    return new Promise((resolve, reject) => {
+      img.onload = async () => {
+        onSelect?.(file)
+        let data
+        const variables = {
+          type: file.type,
+          size: file.size,
+          width: img.width,
+          height: img.height
+        }
+        try {
+          ({ data } = await getSignedPOST({ variables }))
+        } catch (e) {
+          toaster.danger(e.message || e.toString?.())
+          onError?.({ ...variables, name: file.name, file })
+          reject(e)
+          return
+        }
+
+        const form = new FormData()
+        Object.keys(data.getSignedPOST.fields).forEach(key => form.append(key, data.getSignedPOST.fields[key]))
+        form.append('Content-Type', file.type)
+        form.append('Cache-Control', 'max-age=31536000')
+        form.append('acl', 'public-read')
+        form.append('file', file)
+
+        const res = await fetch(data.getSignedPOST.url, {
+          method: 'POST',
+          body: form
+        })
+
+        if (!res.ok) {
+          // TODO make sure this is actually a helpful error message and does not expose anything to the user we don't want
+          const err = res.statusText
+          toaster.danger(err)
+          onError?.({ ...variables, name: file.name, file })
+          reject(err)
+          return
+        }
+
+        const url = `https://${process.env.NEXT_PUBLIC_AWS_UPLOAD_BUCKET}.s3.amazonaws.com/${data.getSignedPOST.fields.key}`
+        // key is upload id in database
+        const id = data.getSignedPOST.fields.key
+        onSuccess?.({ ...variables, id, name: file.name, url, file })
+        resolve(id)
+      }
+    })
+  }, [toaster, getSignedPOST])
+
+  return (
+    <>
+      <input
+        ref={ref}
+        type='file'
+        multiple
+        className='d-none'
+        accept={UPLOAD_TYPES_ALLOW.join(', ')}
+        onChange={async (e) => {
+          const fileList = e.target.files
+          for (const file of Array.from(fileList)) {
+            if (UPLOAD_TYPES_ALLOW.indexOf(file.type) === -1) {
+              toaster.danger(`image must be ${UPLOAD_TYPES_ALLOW.map(t => t.replace('image/', '')).join(', ')}`)
+              continue
+            }
+            await s3Upload(file)
+            // TODO find out if this is needed and if so, why (copied from components/upload.js)
+            e.target.value = null
```

Comment:
I copied this from _components/upload.js_. @huumn, you know what is the purpose of this? You added this in 29fb37b7632960871e41c4055f84b06d8911b997.

I also need to refactor this code with that code anyway.

Reply:
I'm not sure, it might be dead code. Worth testing with/without. If I had to guess, it might be clearing the file field so they can upload another image if they don't like the one they just uploaded.

Reply:
[You were right, it's to clear the file field](https://bobbyhadz.com/blog/react-reset-file-input#reset-a-file-input-in-react)